crates/red_knot_python_semantic/src/types.rs

```diff
@@ -80,29 +80,54 @@ pub fn check_types(db: &dyn Db, file: File) -> TypeCheckDiagnostics {
     diagnostics
 }
 
+/// Computes a possibly-widened type from the inferred type of a symbol.
+///
+/// When accessing a symbol from another scope that has no declared type, we use this to
+/// compute the union `Unknown | T_inferred`, unless the symbol is considered non-modifiable.
+fn widen_type_for_undeclared_public_symbol<'db>(
+    db: &'db dyn Db,
+    inferred: Symbol<'db>,
+    is_considered_non_modifiable: bool,
+) -> Symbol<'db> {
+    // We special-case known-instance types here since symbols like `typing.Any` are typically
+    // not declared in the stubs (e.g. `Any = object()`), but we still want to treat them as
+    // such.
+    let is_known_instance = inferred
+        .ignore_possibly_unbound()
+        .is_some_and(|ty| matches!(ty, Type::KnownInstance(_)));
```

Comment:
This is probably a questionable choice. I don't know what else to do though. The problem appears with `typing.Any`, `typing.List`, `typing.Union`, …, `knot_extensions.Unknown`, `knot_extensions.AlwaysTrue`.